### ISSUE TITLE
v9 Firmware APIs

### ISSUE THREAD

**Issue by timdorr** [timdorr/tesla-api#82]
There are a few new APIs to go along with the v9 firmware release tonight:

- Media controls
- Nav request from your phone (via its Share function)
- Starting a firmware update

There is also a new `energy_sites` endpoint which appears to have a list of Superchargers and their live status info. Exciting stuff!

Let's get it documented!

**Comment by jonasman** [comment]
Wow!!! for the last one i can make an iOS app for everyone to use :) I will try it out today

**Comment by bahree** [comment]
I also see Diagnostics Entitlement and Send Diagnostic endpoint - is that new and related to the car, as opposed to the powerwall? I don't recall seeing that earlier but not 100% sure. Also see a send device key which I wonder is another level of auth or something else.

**Comment by jonasman** [comment]
@timdorr how do you use the new `energy_sites` endpoint? 
a normal GET is not working

**Comment by timdorr** [comment]
Oh, you know what, `energy_sites` is probably for their solar products. `/api/1/energy_sites/{site_id}/storm_mode` kinda gives that away.

You'll need to actually own one, though. Everything you own is available under `/api/1/products`.

**Comment by ronakdesai** [comment]
http://desigeek.com/blog/amit/2018/10/01/tesla-v9-api-endpoints/ is probably helpful

**Comment by timdorr** [comment]
Yep, that's all in endpoints.json. It's just an apktool command away!

**Comment by bahree** [comment]
Are there any changes to the json data structures of the existing calls? I haven't had the time to go and check those. Tim I don't know if you already have some smoke tests that compares?

**Comment by timdorr** [comment]
I plan on updating all the response bodies when updating the docs.

**Comment by mitchcapper** [comment]
Most of the new commands are just url based post commands off of command/ no data is sent with them:
media_toggle_playback / media_next_track / media_prev_track / media_next_fav / media_prev_fav / media_volume_down / media_volume_up

command/navigation_request is more interesting as it seems it is actually not sending the location but rather the actual share data directly to the tesla and they do server side parsing.  For example shares from google maps:
fails:
{"type":"share_ext_content_raw","value":{"android.intent.ACTION":"android.intent.action.SEND","android.intent.TYPE":"text\/plain","android.intent.extra.SUBJECT":"Dropped Pin","android.intent.extra.TEXT":"Dropped Pin\nnear Something\nhttps:\/\/maps.google.com\/?q=1234,1234"},"locale":"en-US","timestamp_ms":"1234.1234"}
Matched multiple addresses (so maybe it expect street then a new line with city/state but as it has street twice got confused:
{"type":"share_ext_content_raw","value":{"android.intent.ACTION":"android.intent.action.SEND","android.intent.TYPE":"text\/plain","android.intent.extra.SUBJECT":"1234 13th st","android.intent.extra.TEXT":"1234 13th st\n1234 13th st, Beverly Hills, WA 90210, USA\nhttps:\/\/maps.google.com\/?q=1234&ftid=0x12345:0x1234"},"locale":"en-US","timestamp_ms":"1234"}

Worked well:
{"type":"share_ext_content_raw","value":{"android.intent.ACTION":"android.intent.action.SEND","android.intent.TYPE":"text\/plain","android.intent.extra.SUBJECT":"Place Name","android.intent.extra.TEXT":"Place Name\n1234 Main St, Beverly HIlls, CA 90210\n(123) 123-1234\nhttps:\/\/maps.google.com\/?cid=12345"},"locale":"en-US","timestamp_ms":"12345"}


It does not use the gps part of the string even with google maps provides is currently (although they could easily change that in the future) but instead seems to rely on the first part of the android.intent.extra.TEXT that looks like an address. It goes beyond just "SUBJECT" as when subject was "Dropped Pin" it still seemed to use the text data. I have not played around too much yet.

**Comment by drewbaumann** [comment]
I was trying to get an idea of what is being sent out by the iOS share option for navigation, but ran into SSL issues going through Charles for iOS. Is there a recommended approach for reverse engineering these calls on iOS?

**Comment by mitchcapper** [comment]
@drewbaumann tesla uses ssl pinning which you must bypass to use a proxy, the other option is to trace the actual raw http calls.  If there are other share forms happy to capture them.   The final one is pretty easy to emulate.  The phone number and google maps link are not used, place name will show up in history.  Just passing the address for the rest.

**Comment by drewbaumann** [comment]
@mitchcapper have you narrowed down the minimum amount of information you need to send for a successful response?

**Comment by JakeShirley** [comment]
FYI If you have a jailbroken iOS device there is a Cydia package to disable SSL pinning.

**Comment by timdorr** [comment]
You can also do so with this Xposed module on Android: https://github.com/ac-pm/Inspeckage

**Comment by mitchcapper** [comment]
The car does use the name of the place as well in history.
